File: backend/src/main/java/com/boke/blog/security/JwtService.java
package com.boke.blog.security;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.security.Keys;
import java.nio.charset.StandardCharsets;
import java.security.Key;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

@Service
public class JwtService {

  private final Key key;
  private final long expireSeconds;

  public JwtService(@Value("${app.jwt.secret}") String secret,
      @Value("${app.jwt.expire-seconds}") long expireSeconds) {
    if (secret.length() >= 32) {
      this.key = Keys.hmacShaKeyFor(secret.getBytes(StandardCharsets.UTF_8));
    } else {
      byte[] padded = (secret + "00000000000000000000000000000000").substring(0, 32)
          .getBytes(StandardCharsets.UTF_8);
      this.key = Keys.hmacShaKeyFor(padded);
    }
    this.expireSeconds = expireSeconds;
  }

  public String issueToken(Long userId, String username, List<Long> roleIds, boolean isAdmin) {
    Instant now = Instant.now();
    return Jwts.builder()
        .subject(username)
        .claim("userId", userId)
        .claim("roleIds", roleIds)
        .claim("isAdmin", isAdmin)
        .issuedAt(Date.from(now))
        .expiration(Date.from(now.plusSeconds(expireSeconds)))
        .signWith(key)
        .compact();
  }

  @SuppressWarnings("unchecked")
  public AuthPrincipal parseToken(String token) {
    Claims claims = Jwts.parser().verifyWith((javax.crypto.SecretKey) key).build()
        .parseSignedClaims(token).getPayload();
    Object roleObject = claims.get("roleIds");
    List<Long> roleIds = new ArrayList<>();
    if (roleObject instanceof List<?> values) {
      for (Object value : values) {
        if (value instanceof Number number) {
          roleIds.add(number.longValue());
        }
      }
    }
    Number userIdNumber = claims.get("userId", Number.class);
    if (userIdNumber == null) {
      throw new IllegalArgumentException("token userId invalid");
    }
    return new AuthPrincipal(
        userIdNumber.longValue(),
        claims.getSubject(),
        roleIds,
        Boolean.TRUE.equals(claims.get("isAdmin", Boolean.class))
    );
  }
}

File: backend/app/jwt_service.py
from datetime import datetime, timedelta, timezone

import jwt

from app.config import JWT_EXPIRE_SECONDS, JWT_SECRET
from app.schemas import AuthPrincipal


def _secret_key() -> str:
    if len(JWT_SECRET) >= 32:
        return JWT_SECRET
    return (JWT_SECRET + "00000000000000000000000000000000")[:32]


class JwtService:
    def issue_token(
        self,
        user_id: int,
        username: str,
        role_ids: list[int],
        is_admin: bool,
    ) -> str:
        now = datetime.now(timezone.utc)
        payload = {
            "sub": username,
            "userId": user_id,
            "roleIds": role_ids,
            "isAdmin": is_admin,
            "iat": now,
            "exp": now + timedelta(seconds=JWT_EXPIRE_SECONDS),
        }
        return jwt.encode(payload, _secret_key(), algorithm="HS256")

    def parse_token(self, token: str) -> AuthPrincipal:
        payload = jwt.decode(token, _secret_key(), algorithms=["HS256"])
        role_ids = []
        for value in payload.get("roleIds") or []:
            if isinstance(value, (int, float)):
                role_ids.append(int(value))
        user_id = payload.get("userId")
        if user_id is None:
            raise ValueError("token userId invalid")
        return AuthPrincipal(
            user_id=int(user_id),
            username=payload.get("sub", ""),
            role_ids=role_ids,
            admin=bool(payload.get("isAdmin")),
        )
